llvm/lib/Support/Mustache.cpp

```diff
@@ -0,0 +1,463 @@
+    }
+    case Token::Type::Variable: {
+      CurrentNode = new (Node) ASTNode(ASTNode::Variable, A, Parent);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::UnescapeVariable: {
+      CurrentNode = new (Node) ASTNode(ASTNode::UnescapeVariable, A, Parent);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::Partial: {
+      CurrentNode = new (Node) ASTNode(ASTNode::Partial, A, Parent);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::SectionOpen: {
+      CurrentNode = new (Node) ASTNode(ASTNode::Section, A, Parent);
+      size_t Start = CurrentPtr;
+      parseMustache(CurrentNode);
+      size_t End = CurrentPtr;
+      SmallString<128> RawBody;
+      if (Start + 1 < End - 1) {
+        for (std::size_t I = Start + 1; I < End - 1; I++)
+          RawBody += Tokens[I].getRawBody();
+      } else if (Start + 1 == End - 1) {
+        RawBody = Tokens[Start].getRawBody();
+      }
+      CurrentNode->setRawBody(RawBody);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::InvertSectionOpen: {
+      CurrentNode = new (Node) ASTNode(ASTNode::InvertSection, A, Parent);
+      size_t Start = CurrentPtr;
+      parseMustache(CurrentNode);
+      size_t End = CurrentPtr;
+      SmallString<128> RawBody;
+      if (Start + 1 < End - 1) {
+        for (size_t Idx = Start + 1; Idx < End - 1; Idx++)
+          RawBody += Tokens[Idx].getRawBody();
+      } else if (Start + 1 == End - 1) {
+        RawBody = Tokens[Start].getRawBody();
+      }
+      CurrentNode->setRawBody(RawBody);
+      Parent->addChild(CurrentNode);
+      break;
+    }
+    case Token::Type::SectionClose:
+      return;
+    default:
+      break;
+    }
+  }
+}
+
+SmallString<128> Template::render(Value Data) {
+  BumpPtrAllocator LocalAllocator;
+  return Tree->render(Data, LocalAllocator, Partials, Lambdas, SectionLambdas,
+                      Escapes);
+}
+
+void Template::registerPartial(StringRef Name, StringRef Partial) {
+  Parser P = Parser(Partial, Allocator);
+  ASTNode *PartialTree = P.parse();
+  Partials[Name] = PartialTree;
+}
+
+void Template::registerLambda(StringRef Name, Lambda L) { Lambdas[Name] = L; }
+
+void Template::registerLambda(StringRef Name, SectionLambda L) {
+  SectionLambdas[Name] = L;
+}
+
+void Template::registerEscape(DenseMap<char, StringRef> E) { Escapes = E; }
+
+Template::Template(StringRef TemplateStr) {
+  Parser P = Parser(TemplateStr, Allocator);
+  Tree = P.parse();
+  // the default behaviour is to escape html entities
+  DenseMap<char, StringRef> HtmlEntities = {{'&', "&amp;"},
+                                            {'<', "&lt;"},
+                                            {'>', "&gt;"},
+                                            {'"', "&quot;"},
+                                            {'\'', "&#39;"}};
+  registerEscape(HtmlEntities);
+}
+
+SmallString<128> printJson(Value &Data) {
```

Comment:
I'm not totally sure it makes sense for a `printJson` function to return a string, instead of writing to an `Ostream`. Maybe rename to `JsonToString` or `toJsonString`?